From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: etil2jz <blanchot.arthur@protonmail.ch>
Date: Sat, 2 Apr 2022 23:29:24 +0200
Subject: [PATCH] Fix NBT pieces overriding a block entity during worldgen
 deadlock

By checking if the world passed into StructureTemplate's placeInWorld
is not a WorldGenRegion, we can bypass the deadlock entirely.
See https://bugs.mojang.com/browse/MC-246262

diff --git a/src/main/java/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java b/src/main/java/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
index 020e436f77f9f2de631727c00ccd41e9293b1f47..25872fc1e827ee1f736c121282585cacec532fb8 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -266,7 +266,11 @@ public class StructureTemplate {
 
                         if (definedstructure_blockinfo.nbt != null) {
                             tileentity = world.getBlockEntity(blockposition2);
-                            Clearable.tryClear(tileentity);
+                            // Paper start - Fix NBT pieces overriding a block entity during worldgen deadlock
+                            if (!(world instanceof net.minecraft.world.level.WorldGenLevel)) {
+                                Clearable.tryClear(tileentity);
+                            }
+                            // Paper end
                             world.setBlock(blockposition2, Blocks.BARRIER.defaultBlockState(), 20);
                         }
 
@@ -379,7 +383,11 @@ public class StructureTemplate {
                         if (pair1.getSecond() != null) {
                             tileentity = world.getBlockEntity(blockposition6);
                             if (tileentity != null) {
-                                tileentity.setChanged();
+                                // Paper start - Fix NBT pieces overriding a block entity during worldgen deadlock
+                                if (!(world instanceof net.minecraft.world.level.WorldGenLevel)) {
+                                    tileentity.setChanged();
+                                }
+                                // Paper end
                             }
                         }
                     }
